diff -uNr snapshot-20000531-orig/conf/postfix-script-nosgid snapshot-20000531/conf/postfix-script-nosgid
--- snapshot-20000531-orig/conf/postfix-script-nosgid	Fri Dec 17 20:12:22 1999
+++ snapshot-20000531/conf/postfix-script-nosgid	Wed Jul  5 20:01:59 2000
@@ -231,6 +231,18 @@
 
 	find corrupt -type f -exec $WARN damaged message: {} \;
 
+	for src in /etc/localtime /etc/services /etc/resolv.conf \
+		/lib/libnss_dns.so.1 /lib/libnss_dns.so.2; do
+	    dst=$queue_directory$src
+	    test -f $src -a -f $dst || continue
+	    cmp -s $src $dst && continue
+	    (
+	      echo WARNING: The file $dst was originally created as a copy of
+	      echo $src. They are now different. If you have updated $src
+	      echo successfully you probably want to copy that update to $dst.
+	    ) | logger -t postfix-script -p mail.warning -s
+	done
+
 	# XXX also: look for weird stuff, weird permissions, etc.
 	;;
 
diff -uNr snapshot-20000531-orig/conf/postfix-script-sgid snapshot-20000531/conf/postfix-script-sgid
--- snapshot-20000531-orig/conf/postfix-script-sgid	Wed Jul  5 20:02:10 2000
+++ snapshot-20000531/conf/postfix-script-sgid	Wed Jul  5 20:03:57 2000
@@ -232,6 +232,18 @@
 
 	find corrupt -type f -exec $WARN damaged message: {} \;
 
+	for src in /etc/localtime /etc/services /etc/resolv.conf \
+		/lib/libnss_dns.so.1 /lib/libnss_dns.so.2; do
+	    dst=$queue_directory$src
+	    test -f $src -a -f $dst || continue
+	    cmp -s $src $dst && continue
+	    (
+	      echo WARNING: The file $dst was originally created as a copy of
+	      echo $src. They are now different. If you have updated $src
+	      echo successfully you probably want to copy that update to $dst.
+	    ) | logger -t postfix-script -p mail.warning -s
+	done
+
 	# XXX also: look for weird stuff, weird permissions, etc.
 	;;
 
