*** snapshot-20011008/src/cleanup/cleanup_map1n.c.orig	Thu Oct 26 12:40:26 2000
--- snapshot-20011008/src/cleanup/cleanup_map1n.c	Mon Oct 29 17:52:40 2001
***************
*** 101,107 ****
  	    break;
  	}
  	for (count = 0; /* void */ ; count++) {
! 	    if (been_here_fixed(been_here, argv->argv[arg]) != 0)
  		break;
  	    if (count >= MAX_RECURSION) {
  		msg_warn("%s: unreasonable %s map nesting for %s",
--- 101,111 ----
  	    break;
  	}
  	for (count = 0; /* void */ ; count++) {
! 
! 	    /*
! 	     * Don't expand an address that already expanded into itself.
! 	     */
! 	    if (been_here_check_fixed(been_here, argv->argv[arg]) != 0)
  		break;
  	    if (count >= MAX_RECURSION) {
  		msg_warn("%s: unreasonable %s map nesting for %s",
***************
*** 118,123 ****
--- 122,133 ----
  			argv_add(argv, STR(state->temp1), ARGV_END);
  			argv_terminate(argv);
  		    }
+ 
+ 		    /*
+ 		     * Allow an address to expand into itself once.
+ 		     */
+ 		    if (strcasecmp(saved_lhs, STR(state->temp1)) == 0)
+ 			been_here_fixed(been_here, argv->argv[arg]);
  		}
  		myfree(saved_lhs);
  		argv_free(lookup);
