--- postfix.2.2.8/src/smtpd/smtpd.c.multiline.orig	2006-02-07 12:18:17.000000000 +0100
+++ postfix.2.2.8/src/smtpd/smtpd.c.multiline	2006-02-07 12:19:56.000000000 +0100
@@ -2006,7 +2006,7 @@
 		    "\t(envelope-from %s)", STR(state->buffer));
 #endif
     }
-    smtpd_chat_reply(state, "354 End data with <CR><LF>.<CR><LF>");
+    smtpd_chat_reply(state, "354 Start mail input; end with <CR><LF>.<CR><LF>");
 
     /*
      * Copy the message content. If the cleanup process has a problem, keep

