--- postfix-1.1.12-20021209/makedefs.orig       Tue Dec 10 16:56:58 2002
+++ postfix-1.1.12-20021209/makedefs    Tue Dec 10 16:58:44 2002
@@ -191,10 +191,11 @@
                    echo "See the RELEASE_NOTES file for more information." 1>&2
                    exit 1
                fi
-               if [ -f /usr/include/pcre/pcre.h ]
-               then
-                   CCARGS="$CCARGS -DHAS_PCRE -I/usr/include/pcre"
-               fi
+# RPM Patch: don't automatically include PCRE support
+#              if [ -f /usr/include/pcre/pcre.h ]
+#              then
+#                  CCARGS="$CCARGS -DHAS_PCRE -I/usr/include/pcre"
+#              fi
                # GDBM locks the DBM .pag file after open. This breaks postmap.
                # if [ -f /usr/include/gdbm-ndbm.h ]
                # then
