*** postfix-1.1.5/conf/sample-smtp.cf-	Mon Dec 10 20:04:51 2001
--- postfix-1.1.5/conf/sample-smtp.cf	Wed Mar 20 13:54:31 2002
***************
*** 73,78 ****
--- 73,89 ----
  # 
  #smtp_break_lines = yes
  
+ # The smtp_helo_name parameter specifies the hostname to send along
+ # in the EHLO or HELO command. 
+ # 
+ # The default value is the machine hostname.  Specify a hostname or
+ # [ip.address].  This can be used in the main.cf file, or in the
+ # master.cf file, for example:
+ # 
+ # 	smtp ... smtp -o smtp_helo_name=foo.bar.com
+ # 
+ #smtp_helo_name = $myhostname
+ 
  # The smtp_skip_4xx_greeting parameter controls what happens when
  # an SMTP server greets us with a 4XX status code (go away, try
  # again later).
*** postfix-1.1.5/src/global/mail_params.h-	Fri Feb  1 08:59:51 2002
--- postfix-1.1.5/src/global/mail_params.h	Wed Mar 20 13:41:53 2002
***************
*** 734,739 ****
--- 734,743 ----
  #define DEF_SMTP_BIND_ADDR	""
  extern char *var_smtp_bind_addr;
  
+ #define VAR_SMTP_HELO_NAME	"smtp_helo_name"
+ #define DEF_SMTP_HELO_NAME	"$myhostname"
+ extern char *var_smtp_helo_name;
+ 
  #define VAR_SMTP_RAND_ADDR	"smtp_randomize_addresses"
  #define DEF_SMTP_RAND_ADDR	1
  extern bool var_smtp_rand_addr;
*** postfix-1.1.5/src/smtp/smtp.c-	Tue Mar  5 20:04:30 2002
--- postfix-1.1.5/src/smtp/smtp.c	Wed Mar 20 13:36:57 2002
***************
*** 100,105 ****
--- 100,107 ----
  /* .IP \fBsmtp_break_lines\fR
  /*	Break lines > \fB$line_length_limit\fR into multiple shorter lines.
  /*	Some SMTP servers misbehave on long lines.
+ /* .IP \fBsmtp_helo_name\fR
+ /*	The hostname to be used in HELO and EHLO commands.
  /* .IP \fBsmtp_skip_4xx_greeting\fR
  /*	Skip servers that greet us with a 4xx status code.
  /* .IP \fBsmtp_skip_5xx_greeting\fR
***************
*** 262,267 ****
--- 264,270 ----
  bool    var_smtp_break_lines;
  int     var_smtp_pix_thresh;
  int     var_smtp_pix_delay;
+ char   *var_smtp_helo_name;
  
   /*
    * Global variables. smtp_errno is set by the address lookup routines and by
***************
*** 413,418 ****
--- 416,422 ----
  	VAR_SMTP_SASL_PASSWD, DEF_SMTP_SASL_PASSWD, &var_smtp_sasl_passwd, 0, 0,
  	VAR_SMTP_SASL_OPTS, DEF_SMTP_SASL_OPTS, &var_smtp_sasl_opts, 0, 0,
  	VAR_SMTP_BIND_ADDR, DEF_SMTP_BIND_ADDR, &var_smtp_bind_addr, 0, 0,
+ 	VAR_SMTP_HELO_NAME, DEF_SMTP_HELO_NAME, &var_smtp_helo_name, 1, 0,
  	0,
      };
      static CONFIG_TIME_TABLE time_table[] = {
*** postfix-1.1.5/src/smtp/smtp_proto.c-	Wed Dec 26 16:21:04 2001
--- postfix-1.1.5/src/smtp/smtp_proto.c	Wed Mar 20 13:51:35 2002
***************
*** 207,218 ****
       * heuristic failed.
       */
      if (state->features & SMTP_FEATURE_ESMTP) {
! 	smtp_chat_cmd(state, "EHLO %s", var_myhostname);
  	if ((resp = smtp_chat_resp(state))->code / 100 != 2)
  	    state->features &= ~SMTP_FEATURE_ESMTP;
      }
      if ((state->features & SMTP_FEATURE_ESMTP) == 0) {
! 	smtp_chat_cmd(state, "HELO %s", var_myhostname);
  	if ((resp = smtp_chat_resp(state))->code / 100 != 2)
  	    return (smtp_site_fail(state, resp->code,
  				   "host %s refused to talk to me: %s",
--- 207,218 ----
       * heuristic failed.
       */
      if (state->features & SMTP_FEATURE_ESMTP) {
! 	smtp_chat_cmd(state, "EHLO %s", var_smtp_helo_name);
  	if ((resp = smtp_chat_resp(state))->code / 100 != 2)
  	    state->features &= ~SMTP_FEATURE_ESMTP;
      }
      if ((state->features & SMTP_FEATURE_ESMTP) == 0) {
! 	smtp_chat_cmd(state, "HELO %s", var_smtp_helo_name);
  	if ((resp = smtp_chat_resp(state))->code / 100 != 2)
  	    return (smtp_site_fail(state, resp->code,
  				   "host %s refused to talk to me: %s",
