--- postfix-2.2-20050206/makedefs.orig	2005-02-09 18:45:35.000000000 +0100
+++ postfix-2.2-20050206/makedefs	2005-02-09 18:46:12.000000000 +0100
@@ -344,13 +344,13 @@
 # /usr/lib, or in their /usr/local equivalents. If the OpenSSL files
 # are in a non-standard place, their locations need to be specified.
 #
-#case "$CCARGS" in
-# *-DUSE_TLS*)	;;
-#  *-DNO_TLS*)	;;
-#           *)	CCARGS="$CCARGS -DUSE_TLS"
-#		AUXLIBS="$AUXLIBS -lssl -lcrypto"
-#		;;
-#esac
+case "$CCARGS" in
+ *-DUSE_TLS*)	;;
+  *-DNO_TLS*)	;;
+           *)	CCARGS="$CCARGS -DUSE_TLS"
+		AUXLIBS="$AUXLIBS -lssl -lcrypto"
+		;;
+esac
 
 #
 # PCRE 3.x has a pcre-config utility so we don't have to guess.
