diff -uNr postfix-19991231-pl04.orig/smtpd/smtpd.c postfix-19991231-pl04/smtpd/smtpd.c
--- postfix-19991231-pl04.orig/smtpd/smtpd.c	Thu Jan  6 15:55:48 2000
+++ postfix-19991231-pl04/smtpd/smtpd.c	Tue Feb  1 19:56:00 2000
@@ -550,9 +550,10 @@
 {
     char   *err;
     int     narg;
-    off_t   size = 0;
     char   *arg;
 
+    state->msg_size = 0;
+
     /*
      * Sanity checks. XXX Ignore bad SIZE= values until we can reliably and
      * portably detect overflows while converting from string to off_t.
@@ -589,8 +590,8 @@
 	    || strcasecmp(arg, "BODY=7BIT") == 0) {
 	     /* void */ ;
 	} else if (strncasecmp(arg, "SIZE=", 5) == 0) {
-	    if ((size = off_cvt_string(arg + 5)) < 0)
-		size = 0;
+	    if ((state->msg_size = off_cvt_string(arg + 5)) < 0)
+		state->msg_size = 0;
 	} else {
 	    state->error_mask |= MAIL_ERROR_PROTOCOL;
 	    smtpd_chat_reply(state, "555 Unsupported option: %s", arg);
@@ -604,7 +605,9 @@
 	smtpd_chat_reply(state, "%s", err);
 	return (-1);
     }
-    if ((err = smtpd_check_size(state, size)) != 0) {
+    if (SMTPD_STAND_ALONE(state) == 0
+	&& var_smtpd_delay_reject == 0
+        && (err = smtpd_check_size(state, state->msg_size)) != 0) {
 	smtpd_chat_reply(state, "%s", err);
 	return (-1);
     }
diff -uNr postfix-19991231-pl06.orig/smtpd/smtpd.h postfix-19991231-pl06/smtpd/smtpd.h
--- postfix-19991231-pl06/smtpd/smtpd.h.orig        Thu Mar 30 22:33:59 2000
+++ postfix-19991231-pl06/smtpd/smtpd.h     Thu Mar 30 22:34:52 2000
@@ -53,6 +53,7 @@
     char   *where;
     int     recursion;
     int     junk_cmds;
+    off_t   msg_size;
 } SMTPD_STATE;

 extern void smtpd_state_init(SMTPD_STATE *, VSTREAM *);
diff -uNr postfix-19991231-pl06.orig/smtpd/smtpd_check.c postfix-19991231-pl06/smtpd/smtpd_check.c
--- postfix-19991231-pl06.orig/smtpd_check.c	Thu Mar 30 22:50:49 2000
+++ postfix-19991231-pl06/smtpd/smtpd_check.c	Thu Mar 30 22:53:04 2000
@@ -1849,7 +1849,9 @@
     if (var_smtpd_delay_reject)
 	if ((err = smtpd_check_client(state)) != 0
 	    || (err = smtpd_check_helo(state, state->helo_name)) != 0
-	    || (err = smtpd_check_mail(state, state->sender)) != 0)
+	    || (err = smtpd_check_mail(state, state->sender)) != 0
+	    || (err = smtpd_check_size(state, state->msg_size)) != 0)
+
 	    SMTPD_CHECK_RCPT_RETURN(err);
 
     /*
diff -uNr postfix-19991231-pl06.orig/smtpd/smtpd_state.c postfix-19991231-pl06/smtpd/smtpd_state.c
--- postfix-19991231-pl06.orig/smtpd/smtpd_state.c	Sun Nov 21 01:42:33 1999
+++ postfix-19991231-pl06/smtpd/smtpd_state.c	Tue Feb  1 19:25:35 2000
@@ -87,6 +87,7 @@
     state->where = SMTPD_AFTER_CONNECT;
     state->recursion = 0;
     state->junk_cmds = 0;
+    state->msg_size = 0;
 
     /*
      * Initialize peer information.
