diff -uNr postfix-2.10.2.orig/src/smtpd/smtpd.c postfix-2.10.2/src/smtpd/smtpd.c
--- postfix-2.10.2.orig/src/smtpd/smtpd.c	2012-12-26 00:58:01.000000000 +0100
+++ postfix-2.10.2/src/smtpd/smtpd.c	2013-12-29 09:37:33.860281012 +0100
@@ -4691,7 +4691,7 @@
 				 var_myhostname);
 		/* Not: state->error_count++; */
 	    } else {
-		smtpd_chat_reply(state, "220 %s", var_smtpd_banner);
+		smtpd_chat_reply_multiline(state, 220, var_smtpd_banner);
 	    }
 	}
 
diff -uNr postfix-2.10.2.orig/src/smtpd/smtpd_chat.c postfix-2.10.2/src/smtpd/smtpd_chat.c
--- postfix-2.10.2.orig/src/smtpd/smtpd_chat.c	2012-01-14 23:31:52.000000000 +0100
+++ postfix-2.10.2/src/smtpd/smtpd_chat.c	2013-12-29 09:41:46.105279869 +0100
@@ -62,6 +62,7 @@
 #include <time.h>
 #include <stdlib.h>			/* 44BSD stdarg.h uses abort() */
 #include <stdarg.h>
+#include <ctype.h>
 
 /* Utility library. */
 
@@ -225,6 +226,61 @@
 	state->flags |= SMTPD_FLAG_HANGUP;
 }
 
+/* trim_line - remove trailing whitespace and return a pointer to the first */
+/* non-space char in the string.  Based on TRIM() in ../util/dict.c         */
+/* WARNING: this function is DESTRUCTIVE.                                   */
+
+static char *trim_line( char *line ) {
+    char *p;
+
+    for (p=line + strlen(line); p > line && ISSPACE(p[-1]); p--);
+	*p = 0;
+
+    while ( ISSPACE(*line) )
+       ++line;
+
+    return ( line );
+}
+
+/* smtpd_chat_reply_multiline - convert multline string into seperate */
+/* lines, sending each one to smtpd_chat_reply()                      */
+
+#define LINE_SEPARATOR "\n"
+
+void smtpd_chat_reply_multiline(SMTPD_STATE *state, int smtp_reply_code, char *format,...)
+{
+    va_list ap;
+    char    *line, *line2;     /* one line (of the multiline) reply */
+    size_t   size;             /* size of a single line             */
+    static VSTRING *temp_line; /* SMTP reply before unescapeing     */
+    static VSTRING *multiline; /* SMTP reply after unescaping       */
+
+    /* First-time intialization. */
+    if (!temp_line)
+       temp_line = vstring_alloc(512);
+    if (!multiline)
+       multiline = vstring_alloc(512);
+
+    va_start(ap, format);
+    vstring_vsprintf(temp_line, format, ap);
+    va_end(ap);
+
+    /* unescape lines with "\\n", converting them to "\n" */
+    unescape( multiline, STR(temp_line) );
+
+    line = STR(multiline);
+    while ( (size = strcspn(line, LINE_SEPARATOR)) < strlen(line) ) {
+       *(line + size) = 0;
+
+       line2 = trim_line(line);
+       smtpd_chat_reply(state, "%03d-%s", smtp_reply_code, line2);
+
+       line += size + 1;
+    }
+    line2 = trim_line(line);
+    smtpd_chat_reply(state, "%03d %s", smtp_reply_code, line2);
+}
+
 /* print_line - line_wrap callback */
 
 static void print_line(const char *str, int len, int indent, char *context)
diff -uNr postfix-2.10.2.orig/src/smtpd/smtpd_chat.h postfix-2.10.2/src/smtpd/smtpd_chat.h
--- postfix-2.10.2.orig/src/smtpd/smtpd_chat.h	2006-04-24 20:48:28.000000000 +0200
+++ postfix-2.10.2/src/smtpd/smtpd_chat.h	2013-12-29 09:42:00.337281183 +0100
@@ -15,6 +15,7 @@
 extern void smtpd_chat_reset(SMTPD_STATE *);
 extern void smtpd_chat_query(SMTPD_STATE *);
 extern void PRINTFLIKE(2, 3) smtpd_chat_reply(SMTPD_STATE *, const char *, ...);
+extern void smtpd_chat_reply_multiline(SMTPD_STATE *, int, char *, ...);
 extern void smtpd_chat_notify(SMTPD_STATE *);
 
 /* LICENSE
