*** snapshot-20001217.orig/src/global/mail_params.h	Sat Dec 16 17:22:27 2000
--- snapshot-20001217/src/global/mail_params.h	Sun Jan 21 19:33:46 2001
***************
*** 726,731 ****
--- 726,735 ----
  #define DEF_SMTPD_SASL_REALM	"$myhostname"
  extern char *var_smtpd_sasl_realm;
  
+ #define VAR_SMTPD_SASL_BROKEN_AUTH	"smtpd_sasl_broken_auth"
+ #define DEF_SMTPD_SASL_BROKEN_AUTH	0
+ extern bool var_smtpd_sasl_broken_auth;
+ 
   /*
    * SASL authentication support, client side.
    */
diff -r --context snapshot-20001217.orig/src/smtp/smtp_proto.c snapshot-20001217/src/smtp/smtp_proto.c
*** snapshot-20001217.orig/src/smtp/smtp_proto.c	Sun Dec  3 01:43:56 2000
--- snapshot-20001217/src/smtp/smtp_proto.c	Sun Jan 21 19:38:00 2001
***************
*** 226,231 ****
--- 226,233 ----
  #ifdef USE_SASL_AUTH
  	    else if (var_smtp_sasl_enable && strcasecmp(word, "AUTH") == 0)
  		smtp_sasl_helo_auth(state, words);
+ 	    else if (var_smtp_sasl_enable && strncasecmp(word, "AUTH=", 5) == 0)
+ 		smtp_sasl_helo_auth(state, word + 5);
  #endif
  	    else if (strcasecmp(word, var_myhostname) == 0) {
  		msg_warn("host %s replied to HELO/EHLO with my own hostname %s",
diff -r --context snapshot-20001217.orig/src/smtpd/smtpd.c snapshot-20001217/src/smtpd/smtpd.c
*** snapshot-20001217.orig/src/smtpd/smtpd.c	Fri Dec 15 17:07:07 2000
--- snapshot-20001217/src/smtpd/smtpd.c	Sun Jan 21 19:33:46 2001
***************
*** 340,345 ****
--- 340,346 ----
  bool    var_allow_untrust_route;
  int     var_smtpd_junk_cmd_limit;
  bool    var_smtpd_sasl_enable;
+ bool	var_smtpd_sasl_broken_auth;
  char   *var_smtpd_sasl_opts;
  char   *var_smtpd_sasl_realm;
  char   *var_filter_xport;
***************
*** 439,446 ****
  	smtpd_chat_reply(state, "250-SIZE");
      smtpd_chat_reply(state, "250-ETRN");
  #ifdef USE_SASL_AUTH
!     if (var_smtpd_sasl_enable)
! 	smtpd_chat_reply(state, "250-AUTH %s", state->sasl_mechanism_list);
  #endif
      smtpd_chat_reply(state, "250 8BITMIME");
      return (0);
--- 440,450 ----
  	smtpd_chat_reply(state, "250-SIZE");
      smtpd_chat_reply(state, "250-ETRN");
  #ifdef USE_SASL_AUTH
!     if (var_smtpd_sasl_enable) {
! 		if (var_smtpd_sasl_broken_auth)
! 			smtpd_chat_reply(state, "250-AUTH=%s", state->sasl_mechanism_list);
! 		smtpd_chat_reply(state, "250-AUTH %s", state->sasl_mechanism_list);
! 	}
  #endif
      smtpd_chat_reply(state, "250 8BITMIME");
      return (0);
***************
*** 1447,1452 ****
--- 1451,1457 ----
  	VAR_DISABLE_VRFY_CMD, DEF_DISABLE_VRFY_CMD, &var_disable_vrfy_cmd,
  	VAR_ALLOW_UNTRUST_ROUTE, DEF_ALLOW_UNTRUST_ROUTE, &var_allow_untrust_route,
  	VAR_SMTPD_SASL_ENABLE, DEF_SMTPD_SASL_ENABLE, &var_smtpd_sasl_enable,
+ 	VAR_SMTPD_SASL_BROKEN_AUTH, DEF_SMTPD_SASL_BROKEN_AUTH, &var_smtpd_sasl_broken_auth,
  	0,
      };
      static CONFIG_STR_TABLE str_table[] = {

--9jxsPFA5p3P2qPhR--

------- End of forwarded message -------
