diff -ru snapshot-20010808/INSTALL.sh snapshot-20010808.new/INSTALL.sh
--- snapshot-20010808/INSTALL.sh	Tue Oct 16 11:24:50 2001
+++ snapshot-20010808.new/INSTALL.sh	Tue Oct 16 11:26:41 2001
@@ -122,6 +122,7 @@
 : ${mail_owner=postfix}
 : ${setgid=no}
 : ${manpages=/usr/local/man}
+: ${rpm_install=no}
 
 # Find out the location of configuration files.
 
@@ -216,24 +217,31 @@
     exit 1
 }
 
-chown root $tempdir/junk >/dev/null 2>&1 || {
-    echo Error: you have no permission to change file ownership. 1>&2
-    exit 1
-}
+# test chown/chgrp ability only if not called by rpm -b[abi]
 
-chown "$mail_owner" $tempdir/junk >/dev/null 2>&1 || {
-    echo Error: $mail_owner needs an entry in the passwd file. 1>&2
-    echo Remember, $mail_owner must have a dedicated user id and group id. 1>&2
-    exit 1
-}
+case $rpm_install in
+yes) ;;
+  *)
+    chown root $tempdir/junk >/dev/null 2>&1 || {
+        echo Error: you have no permission to change file ownership. 1>&2
+        exit 1
+    }
 
-case $setgid in
-no) ;;
- *) chgrp "$setgid" $tempdir/junk >/dev/null 2>&1 || {
-        echo Error: $setgid needs an entry in the group file. 1>&2
-        echo Remember, $setgid must have a dedicated group id. 1>&2
+    chown "$mail_owner" $tempdir/junk >/dev/null 2>&1 || {
+        echo Error: $mail_owner needs an entry in the passwd file. 1>&2
+        echo Remember, $mail_owner must have a dedicated user id and group id. 1>&2
         exit 1
     }
+
+    case $setgid in
+    no) ;;
+     *) chgrp "$setgid" $tempdir/junk >/dev/null 2>&1 || {
+            echo Error: $setgid needs an entry in the group file. 1>&2
+            echo Remember, $setgid must have a dedicated group id. 1>&2
+            exit 1
+        }
+    esac
+    ;;
 esac
 
 rm -f $tempdir/junk
@@ -312,7 +320,7 @@
 
 test -d $QUEUE_DIRECTORY/maildrop || {
     mkdir -p $QUEUE_DIRECTORY/maildrop || exit 1
-    chown $mail_owner $QUEUE_DIRECTORY/maildrop || exit 1
+    test "$rpm_install" != "yes" && ( chown $mail_owner $QUEUE_DIRECTORY/maildrop || exit 1 )
 }
 
 case $setgid in
@@ -322,9 +330,9 @@
     postfix_script=conf/postfix-script-nosgid
     ;;
  *) 
-    chgrp $setgid $COMMAND_DIRECTORY/postdrop || exit 1
+    test "$rpm_install" != "yes" && ( chgrp $setgid $COMMAND_DIRECTORY/postdrop || exit 1 )
     chmod g+s $COMMAND_DIRECTORY/postdrop || exit 1
-    chgrp $setgid $QUEUE_DIRECTORY/maildrop || exit 1
+    test "$rpm_install" != "yes" && ( chgrp $setgid $QUEUE_DIRECTORY/maildrop || exit 1 )
     chmod 1730 $QUEUE_DIRECTORY/maildrop || exit 1
     postfix_script=conf/postfix-script-sgid
     ;;
