diff -uNr snapshot-20000531-orig/conf/master.cf snapshot-20000531/conf/master.cf
--- snapshot-20000531-orig/conf/master.cf	Wed Jul  5 20:14:08 2000
+++ snapshot-20000531/conf/master.cf	Wed Jul  5 20:17:10 2000
@@ -59,19 +59,19 @@
 # service type	private	unpriv	chroot	wakeup	maxproc	command + args
 # 		(yes)	(yes)	(yes)	(never)	(50)
 # ==========================================================================
-smtp	  inet	n	-	n	-	-	smtpd
-pickup	  fifo	n	n	n	60	1	pickup
-cleanup	  unix	-	-	n	-	0	cleanup
-qmgr	  fifo	n	-	n	300	1	qmgr
-#qmgr	  fifo	n	-	n	300	1	nqmgr
-rewrite	  unix	-	-	n	-	-	trivial-rewrite
-bounce	  unix	-	-	n	-	0	bounce
-defer	  unix	-	-	n	-	0	bounce
-smtp	  unix	-	-	n	-	-	smtp
-showq     unix	n	-	n	-	-	showq
-error     unix	-	-	n	-	-	error
+smtp	  inet	n	-	y	-	-	smtpd
+pickup	  fifo	n	n	y	60	1	pickup
+cleanup	  unix	-	-	y	-	0	cleanup
+qmgr	  fifo	n	-	y	300	1	qmgr
+#qmgr	  fifo	n	-	y	300	1	nqmgr
+rewrite	  unix	-	-	y	-	-	trivial-rewrite
+bounce	  unix	-	-	y	-	0	bounce
+defer	  unix	-	-	y	-	0	bounce
+smtp	  unix	-	-	y	-	-	smtp
+showq     unix	n	-	y	-	-	showq
+error     unix	-	-	y	-	-	error
 local	  unix	-	n	n	-	-	local
-lmtp	  unix	-	-	n	-	-	lmtp
+lmtp	  unix	-	-	y	-	-	lmtp
 cyrus	  unix	-	n	n	-	-	pipe
     flags=R user=cyrus argv=/cyrus/bin/deliver -e -m ${extension} ${user}
 uucp	  unix	-	n	n	-	-	pipe
