diff -uNr postfix-19990317-orig/conf/main.cf postfix-19990317/conf/main.cf
--- postfix-19990317-orig/conf/main.cf	Wed Mar 17 17:35:48 1999
+++ postfix-19990317/conf/main.cf	Thu Mar 18 10:41:27 1999
@@ -179,7 +179,7 @@
 # "postfix reload" to eliminate the delay.
 #
 #alias_maps = dbm:/etc/aliases
-#alias_maps = hash:/etc/aliases
+alias_maps = hash:/etc/aliases
 #alias_maps = hash:/etc/aliases, nis:mail.aliases
 
 # The alias_database parameter specifies the alias database(s) that
