*** snapshot-20010808/src/bounce/bounce_notify_util.c-	Mon Jul  9 10:23:49 2001
--- snapshot-20010808/src/bounce/bounce_notify_util.c	Wed Aug 22 16:38:51 2001
***************
*** 547,553 ****
  	if (var_bounce_limit == 0 || bounce_length < var_bounce_limit) {
  	    bounce_length += VSTRING_LEN(bounce_info->buf) + 2;
  	    status = (REC_PUT_BUF(bounce, rec_type, bounce_info->buf) != rec_type);
! 	}
      }
  
      /*
--- 547,554 ----
  	if (var_bounce_limit == 0 || bounce_length < var_bounce_limit) {
  	    bounce_length += VSTRING_LEN(bounce_info->buf) + 2;
  	    status = (REC_PUT_BUF(bounce, rec_type, bounce_info->buf) != rec_type);
! 	} else
! 	    break;
      }
  
      /*
