--- postfix-2.0.15/src/util/dict_ldap.c.orig	Thu Oct 17 02:26:41 2002
+++ postfix-2.0.15/src/util/dict_ldap.c	Thu Sep  4 12:16:32 2003
@@ -152,9 +152,12 @@
     char   *bind_dn;
     char   *bind_pw;
     int     timeout;
+#ifdef HAS_LDAP_CACHE
+    /* The LDAP_CACHE feature has been depreciated */
     int     cache;
     long    cache_expiry;
     long    cache_size;
+#endif
     int     dereference;
     int     chase_referrals;
     int     debuglevel;
@@ -374,6 +377,7 @@
     /*
      * Set up client-side caching if it's configured.
      */
+#ifdef HAS_LDAP_CACHE
     if (dict_ldap->cache) {
 	if (msg_verbose)
 	    msg_info
@@ -419,6 +423,7 @@
     if (msg_verbose)
 	msg_info("%s: Cached connection handle for LDAP source %s",
 		 myname, dict_ldap->ldapsource);
+#endif
 
     return (0);
 }
@@ -1036,6 +1041,7 @@
 	msg_info("%s: %s is %s", myname, vstring_str(config_param),
 		 dict_ldap->bind_pw);
 
+#ifdef HAS_LDAP_CACHE
     /*
      * get configured value of "ldapsource_cache"; default to false
      */
@@ -1065,6 +1071,7 @@
     if (msg_verbose)
 	msg_info("%s: %s is %ld", myname, vstring_str(config_param),
 		 dict_ldap->cache_size);
+#endif
 
     /*
      * Alias dereferencing suggested by Mike Mattice.
